llvm/lib/Target/AArch64/AArch64FastISel.cpp

```diff
@@ -1233,7 +1233,11 @@ unsigned AArch64FastISel::emitAddSub(bool UseAdd, MVT RetVT, const Value *LHS,
       isValueAvailable(RHS)) {
     if (const auto *SI = dyn_cast<BinaryOperator>(RHS))
       if (const auto *C = dyn_cast<ConstantInt>(SI->getOperand(1)))
-        if ((SI->getOpcode() == Instruction::Shl) && (C->getZExtValue() < 4)) {
+        if ((SI->getOpcode() == Instruction::Shl) && (C->getZExtValue() < 4) &&
+            !NeedExtend) {
+          // We can only fold instructions that doesn't need extension because
```

Comment:
Is it possible to reach this point now, if NeedExtend==false and ExtendType != AArch64_AM::InvalidShiftExtend?

Reply:
I didn't add `ExtendType != AArch64_AM::InvalidShiftExtend` because there is already a check before this nested-if statement in line 1232. Do you think it's still better to add the check?

Reply:
Yeah - I meant that if `ExtendType != InvalidShift` then does that mean that NeedExtend is always false? Is it ever possible to get to the emitAddSub_rx(..) line now?

Reply:
Ah Gotcha. You're right, if `ExtendType != AArch64_AM::InvalidShiftExtend`, NeedExtend is always true. So the logic would never be reached. Maybe it's a better idea to remove that part of the logic all together?